src/amber/router/params.cr

```diff
@@ -65,11 +65,12 @@ module Amber::Router
     end
 
     def to_h
-      query.to_h
-           .merge(form.to_h)
-           .merge(route)
-           .merge(json)
-           .merge(multipart)
+      params_hash = Hash(String, String?).new
+      query.each { |key, _| params_hash[key] = query[key] }
+      route.each { |key, _| params_hash[key] = route[key] }
+      multipart.each { |key, _| params_hash[key] = multipart[key] }
+      form.each { |key, _| params_hash[key] = form[key] }
+      params_hash
```

Comment:
This should be equivalent but a bit faster/cleaner:

```crystal
query.to_h
     .dup
     .merge(route)
     .merge(multipart)
     .merge(form)
```

Reply:
@veelenga that's essentially the exact code that is causing the bug

Reply:
Nope. It has an additional `dup`.

Reply:
This partially works. The problem we are experiencing besides the mutation is that the value return for a given key is the last value present. If you were to run the following. 

```
query.to_h
     .dup
     .merge(route)
     .merge(multipart)
     .merge(form.to_h)
```

You get the following

```bash
Amber::Router::Params #to_h returns the first value for key
     Failure/Error: url_encoded_form_post.to_h["cat"].should eq "1"

       Expected: "1"
            got: "-1"
```

Reply:
@eliasjpr Why not invert the order then? :sweat_smile: 

```
form.to_h
    .dup
    .merge(multipart)
    .merge(route)
    .merge(query.to_h)
```

Reply:
The bug here has nothing to do with the way params from different fields are merged into the params hash. It has to do with the way multiple values from the same hash are chosen.

In other words, given this series of form parameters:

```html
<input name="animal" value="cat">
<input name="animal" value="dog">
```

What is the value of `params["animal"]`?
What is the value of `params.raw_params["animal"]`? 
What is the value of `params.raw_params.to_h["animal"]`?

All three should be the same. With the code in master, you get:

`params["animal"] => "cat"`
`params.raw_params["animal"] => "cat"`
`params.raw_params.to_h["animal"] => "dog"`

This is the problem that running a straight `.merge` creates, and why it does not work.